url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/gui/quantum/Entangler.java

```diff
@@ -0,0 +1,212 @@
+/*******************************************************************************
+ * Copyright (c) 2013 Martin Marinov.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the GNU Public License v3.0
+ * which accompanies this distribution, and is available at
+ * http://www.gnu.org/licenses/gpl.html
+ * 
+ * Contributors:
+ *     Martin - initial API and implementation
+ ******************************************************************************/
+package martin.gui.quantum;
+
+import java.awt.BasicStroke;
+import java.awt.Graphics2D;
+import java.awt.Stroke;
+import java.util.ArrayList;
+import java.util.HashSet;
+
+
+public class Entangler extends Item {
+	
+	private final String[] menu_entries = new String[] {};
+	
+	public static final int QUBIT_SIZE = 32;
+	
+	private static final int ICON_OFFSET = 16;
+	private final static Stroke DASHED_STROKE = new BasicStroke(4.0f);
+	private static final int PICKING_DISTANCE = 6;
+	
+	Qubit i1, i2;
+	
+	private boolean select_first = true;
+	
+	public Entangler() {}
+	
+	private Entangler(Qubit i1, Qubit i2) {
+		this.i1 = i1;
+		this.i2 = i2;
+	}
+	
+	@Override
+	public boolean equals(Object obj) {
+		if (obj instanceof Entangler) {
+			final Entangler o = (Entangler) obj;
+			return (o.i1 == i1 && o.i2 == i2) || (o.i1 == i2 && o.i2 == i1);
+		}
+		return false;
+	}
+
+	@Override
+	void renderInstance(Graphics2D g, Visualizer vis) {
+		final int dx = i2.x - i1.x;
+		final int dy = i2.y - i1.y;
+		final double d = Math.sqrt(dx * dx + dy * dy);
+		final double cos = dx / d;
+		final double sin = dy / d;
+
+		final int sx = i1.x + (int) (cos * QUBIT_SIZE);
+		final int sy = i1.y + (int) (sin * QUBIT_SIZE);
+		final int ex = i2.x - (int) (cos * QUBIT_SIZE);
+		final int ey = i2.y - (int) (sin * QUBIT_SIZE);
+		
+		final Stroke def = g.getStroke();
+		g.setStroke(DASHED_STROKE);
+		g.drawLine(sx, sy, ex, ey);
+		g.setStroke(def);
+	}
+
+	@Override
+	boolean isMouseOntop(int x, int y) {
+		final double dist = distFromLine(x, y);
+		
+		if (dist < PICKING_DISTANCE)
+			return true;
+		
+		return false;
+	}
+	
+	public double distFromLine(int px, int py) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 87-104 into a separate method.